libsolidity/analysis/DeclarationTypeChecker.cpp

```diff
@@ -78,26 +78,61 @@ bool DeclarationTypeChecker::visit(StructDefinition const& _struct)
 	}
 
 	bool previousRecursiveStructSeen = m_recursiveStructSeen;
+	bool previousHasNestedMapping = m_nestedMappingSeen;
 	bool hasRecursiveChild = false;
+	bool hasNestedMapping = false;
 
 	m_currentStructsSeen.insert(&_struct);
 
 	for (auto const& member: _struct.members())
 	{
 		m_recursiveStructSeen = false;
+		m_nestedMappingSeen = false;
 		member->accept(*this);
 		solAssert(member->annotation().type, "");
 		solAssert(member->annotation().type->canBeStored(), "Type cannot be used in struct.");
+
+		if (member->type()->category() == Type::Category::Mapping)
```

Comment:
I'm wondering if we actually need this or if we can just use `m_nestedMappingSeen` and do this in the individual visitor functions. Did you try that?

Reply:
I think we can replace the `if (member->type()->category() == Type::Category::Mapping)` and set a variable in `DeclarationTypeChecker` whenever mappings are visited. This will have to be unset at the beginning of `visit(StructDefinition...)`. 

 But the conditions 
1. `else if (auto structType = dynamic_cast<StructType const*>(finalBaseType))` 
2. `else if (auto structType = dynamic_cast<StructType const*>(member->type()))`

will have to remain. Because, if those  structs were visited before, it's members won't be visited again. So in the end, maybe doing it like this is the cleanest way, I think.

Did you have something in mind that will simply the checks?

Reply:
I just think this gets really complicated. Do you think it would be possible to disentangle this from the recursion checks, i.e. check for nested mappings in its own component?

Reply:
I moved it here: https://github.com/ethereum/solidity/pull/9146/files#diff-07335c711d9742b9c53a75648c7ad55cR2269